_posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd

```diff
@@ -0,0 +1,390 @@
+
+#### Define model
+
+Our example model is a straightforward multi-layer perceptron. Note, though, that it has two dropout layers. Dropout layers behave differently during training and evaluation; and as we've learned, decisions made during tracing are set in stone. This is something we'll need to take care of once we're done training the model.
+
+```{r}
+library(torch)
+net <- nn_module( 
+  
+  initialize = function() {
+    
+    self$l1 <- nn_linear(3, 8)
+    self$l2 <- nn_linear(8, 16)
+    self$l3 <- nn_linear(16, 1)
+    self$d1 <- nn_dropout(0.2)
+    self$d2 <- nn_dropout(0.2)
+    
+  },
+  
+  forward = function(x) {
+    x %>%
+      self$l1() %>%
+      nnf_relu() %>%
+      self$d1() %>%
+      self$l2() %>%
+      nnf_relu() %>%
+      self$d2() %>%
+      self$l3()
+  }
+)
+
+train_model <- net()
+```
+
+#### Train model on toy dataset
+
+For demonstration purposes, we create a toy dataset with three predictors and a scalar target.
+
+```{r}
+toy_dataset <- dataset(
+  
+  name = "toy_dataset",
+  
+  initialize = function(input_dim, n) {
+    
+    df <- na.omit(df) 
+    self$x <- torch_randn(n, input_dim)
+    self$y <- self$x[, 1, drop = FALSE] * 0.2 -
+      self$x[, 2, drop = FALSE] * 1.3 -
+      self$x[, 3, drop = FALSE] * 0.5 +
+      torch_randn(n, 1)
+    
+  },
+  
+  .getitem = function(i) {
+    list(x = self$x[i, ], y = self$y[i])
+  },
+  
+  .length = function() {
+    self$x$size()[[1]]
+  }
+)
+
+input_dim <- 3
+n <- 1000
+
+train_ds <- toy_dataset(input_dim, n)
+
+train_dl <- dataloader(train_ds, shuffle = TRUE)
+
+```
+
+We train long enough to make sure we can distinguish an untrained model's output from that of a trained one.
+
+```{r}
+optimizer <- optim_adam(train_model$parameters, lr = 0.001)
+num_epochs <- 10
+
+train_batch <- function(b) {
+  
+  optimizer$zero_grad()
+  output <- train_model(b$x)
+  target <- b$y
+  
+  loss <- nnf_mse_loss(output, target)
+  loss$backward()
+  optimizer$step()
+  
+  loss$item()
+}
+
+for (epoch in 1:num_epochs) {
+  
+  train_loss <- c()
+  
+  coro::loop(for (b in train_dl) {
+    loss <- train_batch(b)
+    train_loss <- c(train_loss, loss)
+  })
+  
+  cat(sprintf("\nEpoch: %d, loss: %3.4f\n", epoch, mean(train_loss)))
+  
+}
+
+```
+
+    Epoch: 1, loss: 2.6753
+
+    Epoch: 2, loss: 1.5629
+
+    Epoch: 3, loss: 1.4295
+
+    Epoch: 4, loss: 1.4170
+
+    Epoch: 5, loss: 1.4007
+
+    Epoch: 6, loss: 1.2775
+
+    Epoch: 7, loss: 1.2971
+
+    Epoch: 8, loss: 1.2499
+
+    Epoch: 9, loss: 1.2824
+
+    Epoch: 10, loss: 1.2596
+
+#### Trace in `eval` mode
+
+Now, for deployment, we want a model that has the same weights as the model we just trained, but does *not* drop out any tensor elements. A way to achieve is to instantiate a fresh model instance, immediately put it into `eval()` mode, trace it, and *then only* endow it with the trained model's weights.
+
+```{r}
+deploy_model <- net()$eval()
```

Comment:
Since you are not training a traced model, I think you could simply do `train_model$eval()` and then `jit_trace(train_model, ...)`. No need to load the weights again or trace a clean `net()`. You would only need that if `train_model()` was also traced becaus in that case `train_model$eval()` would have no effect in disabling the dropout.

Reply:
oh, right, thanks! I think I got really confused on that one ;-)))